sbtc/src/idpack/codec/strategies/bitset.rs

```diff
@@ -0,0 +1,293 @@
+        // We just ensured the values are non-empty, so this is safe.
+        let min_value = values[0];
+        let max_value = values[values.len() - 1];
+
+        // Calculate bitmap size requirements
+        let range = max_value - min_value;
+        let bytes_needed = range.div_ceil(8) as usize;
+
+        // Safety check to prevent OOM for extremely sparse data
+        if bytes_needed > ALLOC_BYTES_LIMIT as usize {
+            return None;
+        }
+
+        // For bitmaps > 7 bytes, we need an explicit length byte
+        Some(bytes_needed + (bytes_needed > 7) as usize)
+    }
+
+    /// Encodes a segment using the BitSet strategy.
+    ///
+    /// The encoding process uses one of three approaches for maximum
+    /// compression:
+    /// 1. Embedded length: For small bitmaps (≤7 bytes), the bitmap length is
+    ///    embedded in the flags byte to save 1 byte
+    /// 2. Standard bitmap: For larger ranges, an explicit length byte plus the
+    ///    minimum number of bitmap bytes needed
+    ///
+    /// Values are encoded as bits relative to the segment's offset, with
+    /// position 0 representing (offset+1).
+    ///
+    /// ## Parameters
+    /// * `flags` - The encoding flags
+    /// * `segment` - The segment to encode
+    /// * `result` - Vector to append encoded bytes to
+    ///
+    /// ## Returns
+    /// Ok(()) on success, or an error if encoding fails
+    fn encode(
+        &self,
+        flags: u8,
+        segment: &Segment,
+        result: &mut Vec<u8>,
+    ) -> Result<(), SegmentEncodeError> {
+        // Calculate bitmap size requirements.
+        let range = segment.range();
+        let bytes_needed = range.div_ceil(8);
+
+        // Safety check to prevent OOM for extremely sparse data
+        if bytes_needed > ALLOC_BYTES_LIMIT as u64 {
+            return Err(SegmentEncodeError::ByteAllocationLimit(bytes_needed));
+        }
+
+        // Safety check to prevent OOM for extremely large bitmaps
+        let value_count = segment.len() as u64;
+        if value_count > VALUE_COUNT_LIMIT as u64 {
+            return Err(SegmentEncodeError::TooManyValues(value_count));
+        }
+
+        // Allocate bitmap array filled with zeros.
+        let mut bitmap = vec![0u8; bytes_needed as usize];
+
+        // Populate the bitmap by setting bits for each value (excluding the
+        // offset).
+        for &value in segment.values() {
+            // Convert from value to bit position:
+            // 1. Subtract offset to get relative position
+            // 2. Subtract 1 more because bit 0 represents (offset+1)
+            let relative_pos = value - segment.offset() - 1;
+
+            // Calculate byte and bit index within the bitmap
+            let byte_index = relative_pos / 8;
+            let bit_index = relative_pos % 8;
+
+            // Set the corresponding bit in the bitmap
+            bitmap[byte_index as usize] |= 1 << bit_index;
```

Comment:
It's not obvious that this is always okay, so let's shy away from the paniking API and use `get_mut` or something.